File: backoffice/src/product/application/ProductCreator.py
"""
 *
 * Libraries 
 *
"""

from src.product.domain    import ProductName
from src.product.domain    import ProductPrice
from src.product.domain    import ProductDescription
from src.product.domain    import ProductId
from src.product.domain    import ProductRepository
from src.product.domain    import Product
from src.restaurant.domain import RestaurantRepository
from src.restaurant.domain import RestaurantId
from src.restaurant.domain import RestaurantNotFoundException
from src.shared.domain     import ServerInternalErrorException
from src.product.domain    import ProductNameAlreadyCreatedException
from src.restaurant.domain import Restaurant
from src.shared.domain     import DomainEventsPublisher

"""
 *
 * Class 
 *
"""

class ProductCreator:

    """
     *
     * Parameters 
     *
    """
 
    __repository           : ProductRepository
    __restaurantRepository : RestaurantRepository
    __eventPublisher       : DomainEventsPublisher

    """
     *
     * Methods 
     *
    """

    def __init__( 
        self, 
        repository           : ProductRepository,
        restaurantRepository : RestaurantRepository,
        eventPublisher       : DomainEventsPublisher
    ) -> None:
        self.__repository           = repository
        self.__restaurantRepository = restaurantRepository
        self.__eventPublisher       = eventPublisher

    def create( 
        self, 
        id           : ProductId, 
        name         : ProductName,
        price        : ProductPrice,
        description  : ProductDescription,
        restaurantId : RestaurantId
    ) -> None:
        # Variables
        product    : Product
        restaurant : Restaurant
        # Code
        restaurant = self.__restaurantRepository.findById( restaurantId )
        if restaurant is None:
            raise RestaurantNotFoundException( restaurantId )
        product = self.__repository.findByNameAndRestaurant( name, restaurantId )
        if product is not None:
            raise ProductNameAlreadyCreatedException( name )
        product = Product.create( id, name, price, description, restaurantId )     
        if not self.__repository.save( product ):
            raise ServerInternalErrorException()
        self.__eventPublisher.publish( product.pullEvents() )

File: preparation/src/product/application/ProductCreator.cs
/* 
 * 
 * Libraries
 *
*/

using preparation.src.product.domain;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.product.application {

    public class ProductCreator : DomainEventSubscriber<ProductCreated> {

        /* 
         * 
         * Attributes
         *
        */

        private ProductRepository repository;

        /* 
         * 
         * Methods
         *
        */

        public ProductCreator( ProductRepository repository ) {
            this.repository = repository;
        }

        public async Task handleEvent( ProductCreated domainEvent ) {
            // Variables
            Product? product;
            // Code
            await Task.Run( () => {
                product = repository.findById( domainEvent.getId() );
                if( product == null ) {
                    product = Product.create(
                        id   : domainEvent.getId(),
                        name : domainEvent.getName()
                    );
                    if( !repository.save( product ) ) {
                        throw new ServerInternalErrorException();
                    }
                }
            } );
        }

    }

}